toolchain/parse/handle_impl.cpp

```diff
@@ -0,0 +1,61 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parse/context.h"
+
+namespace Carbon::Parse {
+
+auto HandleImplAfterIntroducer(Carbon::Parse::Context& context) -> void {
+  auto state = context.PopState();
+  state.state = State::DeclOrDefinitionAsImpl;
+  context.PushState(state);
+
+  switch (context.PositionKind()) {
+    case Lex::TokenKind::Forall:
+      context.PushState(State::ImplAfterForall);
+      context.ConsumeAndDiscard();
+      context.PushState(State::ParamListAsImplicit);
+      break;
+
+    case Lex::TokenKind::As:
+      context.AddLeafNode(NodeKind::ImplAs, context.Consume());
+      context.PushState(State::Expr);
+      break;
+
+    default:
+      context.PushState(State::ImplAs);
+      context.PushStateForExpr(PrecedenceGroup::ForImplAs());
+      break;
+  }
+}
+
+auto HandleImplAfterForall(Carbon::Parse::Context& context) -> void {
+  auto state = context.PopState();
+  if (state.has_error) {
+    context.ReturnErrorOnState();
+    return;
```

Comment:
In this state, you've encountered a `forall` token. Why do you return here instead of letting the `AddNode` call occur, adding a parse node for the error token?

(note too, with the `return`, using `state.has_error` below is equivalent to `/*has_error=*/false`, which might be clearer)

Reply:
I'd meant just dropping the return.

I'd expected though that ParamListAsImplicit would always add a node, even in error cases. That may be a consistency issue. You adding an InvalidParse node is an unintended workaround.

Anyways, current state is fine, though it's probably worth someone (maybe me) changing ParamList behavior there.

Regarding the rest, jotted down some notes: https://docs.google.com/document/d/1RRYMm42osyqhI2LyjrjockYCutQ5dOf8Abu50kTrkX0/edit?resourcekey=0-kHyqOESbOHmzZphUbtLrTw#heading=h.f0xq7he13pm

Reply:
Based on the observed behavior in the test, it seems that ParamListAsImplicit sets error == true on the parent state if it did not output a node. If ParamListAsImplicit does output a node, that node has the error state set to true, but the parent state is left alone.